[CalDAV-personal] Switch item ON only during server reload, but on time for OFF

Dear all,

I observed a behaviour for CalDAV-personal I cannot explain and it is really disturbing.
I created a discussion here: https://community.openhab.org/t/caldav-personal-switch-item-is-always-delayed-for-on-on-time-for-off/60964.

Edit: Same issue is reported here:
https://community.openhab.org/t/caldav-peronal-event-sheduled-only-at-reload/16320/2
and here:
https://community.openhab.org/t/delayed-caldav-execution/34969

## Expected Behavior
The switch associated with a CalDAV-personal calendar should fire on time both for end and beginning of an event

## Current Behavior
If I got it right, the binding provides the ability to set a switch ON at the beginning of the event and OFF at its end. This does not work for me. The scheduler seems to create the jobs correctly, but only the end of the event fires the switch on time.

**However, the trigger is fired when the `reloadInterval` for checking the server occurs.**

Example:
I scheduled an event to switch on a power plug at 02:00h and to switch it off at 03:00h. However, the switch is only activated at 2:22h, but deacitvated on time at 03:00h exactly.

## Possible Solution
The scheduler does not seem to trigger the Switch item. The trigger is only fired when the `reloadInterval` for checking the server occurs.

## Steps to Reproduce (for bugs)
1. Create a CalDAV calender and filll it with events
2. Create a Switch associated with that calendar
3. Wait for switch going to ON (delayed)
4. Wait for switch going to OFF (on time)

## Context
I try to switch a plug for my TV recorder depending on the TV program in my calendar

## Your Environment
* Version used: OpenHAB, milestone, currently 2.4 release
* Raspberry Pi 3B+, running OpenHABian
* Edit: Also same behaviour on Linux Xubuntu, OH 2.5 no. 1491

@9037568, @MathiasMoog 
I tested the code: Removed the standard binding, cleared cache and tmp directories and copied the two *.jar files to the addon-folder.

Unfortunately, I receive errors:
`2019-01-22 20:56:18.306 [INFO ] [arthome.ui.paper.internal.PaperUIApp] - Stopped Paper UI

2019-01-22 20:56:18.344 [INFO ] [panel.internal.HABPanelDashboardTile] - Stopped HABPanel

2019-01-22 20:56:18.362 [INFO ] [ebuilder.internal.HomeBuilderServlet] - Stopped Home Builder

2019-01-22 20:56:18.389 [INFO ] [.dashboard.internal.DashboardService] - Stopped Dashboard

2019-01-22 20:58:59.250 [WARN ] [org.apache.felix.fileinstall        ] - Error while starting bundle: file:/usr/share/openhab2/addons/org.openhab.io.caldav-1.14.0-SNAPSHOT.jar

org.osgi.framework.BundleException: Could not resolve module: org.openhab.io.caldav [192]

  Unresolved requirement: Import-Package: org.apache.commons.codec

	at org.eclipse.osgi.container.Module.start(Module.java:444) ~[?:?]

	at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:383) ~[?:?]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1260) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1233) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startAllBundles(DirectoryWatcher.java:1221) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:515) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:365) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:316) [10:org.apache.felix.fileinstall:3.6.4]

2019-01-22 20:58:59.287 [WARN ] [org.apache.felix.fileinstall        ] - Error while starting bundle: file:/usr/share/openhab2/addons/org.openhab.binding.caldav-personal-1.14.0-SNAPSHOT.jar

org.osgi.framework.BundleException: Could not resolve module: org.openhab.binding.caldav-personal [191]

  Unresolved requirement: Import-Package: org.openhab.core.binding

	at org.eclipse.osgi.container.Module.start(Module.java:444) ~[?:?]

	at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:383) ~[?:?]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1260) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1233) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:520) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:365) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:316) [10:org.apache.felix.fileinstall:3.6.4]

2019-01-22 20:58:59.304 [WARN ] [org.apache.felix.fileinstall        ] - Error while starting bundle: file:/usr/share/openhab2/addons/org.openhab.io.caldav-1.14.0-SNAPSHOT.jar

org.osgi.framework.BundleException: Could not resolve module: org.openhab.io.caldav [192]

  Unresolved requirement: Import-Package: org.apache.commons.codec

	at org.eclipse.osgi.container.Module.start(Module.java:444) ~[?:?]

	at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:383) ~[?:?]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1260) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1233) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:520) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:365) [10:org.apache.felix.fileinstall:3.6.4]

	at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:316) [10:org.apache.felix.fileinstall:3.6.4]

2019-01-22 20:59:07.924 [INFO ] [.core.internal.i18n.I18nProviderImpl] - Time zone set to 'Europe/Berlin'.

2019-01-22 20:59:07.942 [INFO ] [.core.internal.i18n.I18nProviderImpl] - Locale set to 'de_DE'.

2019-01-22 20:59:08.986 [INFO ] [.dashboard.internal.DashboardService] - Started Dashboard at http://192.168.0.18:8080

2019-01-22 20:59:08.991 [INFO ] [.dashboard.internal.DashboardService] - Started Dashboard at https://192.168.0.18:8443

2019-01-22 20:59:25.262 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'default.rules'

2019-01-22 20:59:29.556 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(72)] : doFindMethod: Suitable but non-accessible method found in class org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl

2019-01-22 20:59:29.561 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(72)] : findMethod: Suitable but non-accessible method setScheduler found in class org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl, subclass of org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl

2019-01-22 20:59:29.566 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(72)] : bind method [setScheduler] not found; Component will fail

2019-01-22 20:59:29.569 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(72)] : bind method [setScheduler] not found

2019-01-22 20:59:29.574 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(72)] : doFindMethod: Suitable but non-accessible method found in class org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl

2019-01-22 20:59:29.577 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(72)] : findMethod: Suitable but non-accessible method unsetScheduler found in class org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl, subclass of org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl

2019-01-22 20:59:29.580 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(72)] : unbind method [unsetScheduler] not found; Component will fail

2019-01-22 20:59:29.583 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(72)] : unbind method [unsetScheduler] not found`

Dear all,

@hakan42, @9037568 : **I re-tested the fix**. Although I could swear that I didn't change anything (I've only been working with the Z-Wave network in the last few days), and I get error messages at the beginning (see below), **the switch triggers on time**.

`
2019-01-28 06:31:34.570 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(107)] : doFindMethod: Suitable but non-accessible method found in class org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl
2019-01-28 06:31:34.576 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(107)] : findMethod: Suitable but non-accessible method setScheduler found in class org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl, subclass of org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl
2019-01-28 06:31:34.582 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(107)] : bind method [setScheduler] not found; Component will fail
2019-01-28 06:31:34.585 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(107)] : bind method [setScheduler] not found
2019-01-28 06:31:34.589 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(107)] : doFindMethod: Suitable but non-accessible method found in class org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl
2019-01-28 06:31:34.594 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(107)] : findMethod: Suitable but non-accessible method unsetScheduler found in class org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl, subclass of org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl
2019-01-28 06:31:34.596 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(107)] : unbind method [unsetScheduler] not found; Component will fail
2019-01-28 06:31:34.599 [ERROR] [org.eclipse.smarthome.core          ] - bundle org.eclipse.smarthome.core:0.11.0.oh250M1 (102)[org.eclipse.smarthome.core.internal.scheduler.PeriodicSchedulerImpl(107)] : unbind method [unsetScheduler] not found
...
...
019-01-29 15:55:00.003 [INFO ] [nhab.io.caldav.internal.job.EventJob] - event BEGIN for: 316c740f-8a28-4fb2-b1e8-5270d413d258.kalender.gmx.net(Aufnahme "Verrückt nach Meer"@29.01.2019/15:55-29.01.2019/17:15)
2019-01-29 15:55:00.004 [INFO ] [nhab.io.caldav.internal.job.EventJob] - event BEGIN for: 316c740f-8a28-4fb2-b1e8-5270d413d258.kalender.gmx.net(Aufnahme "Verrückt nach Meer"@30.01.2019/15:55-30.01.2019/17:15)
2019-01-29 15:55:00.008 [WARN ] [ldav_personal.internal.CalDavBinding] - event is before now; ignoring and updating anyway.
2019-01-29 15:55:00.236 [INFO ] [smarthome.model.script.Regel Horizon] - Calendar event switched Horizon power plug ON: Aufnahme "Verrückt nach Meer"
2019-01-29 16:05:02.399 [INFO ] [ome.model.script.Regel Entertainment] - TV set status to on!
2019-01-29 17:15:00.003 [INFO ] [nhab.io.caldav.internal.job.EventJob] - event END for: 316c740f-8a28-4fb2-b1e8-5270d413d258.kalender.gmx.net(Aufnahme "Verrückt nach Meer"@29.01.2019/15:55-29.01.2019/17:15)
2019-01-29 17:15:00.004 [INFO ] [nhab.io.caldav.internal.job.EventJob] - event END for: 316c740f-8a28-4fb2-b1e8-5270d413d258.kalender.gmx.net(Aufnahme "Verrückt nach Meer"@30.01.2019/15:55-30.01.2019/17:15)
2019-01-29 17:15:00.155 [INFO ] [smarthome.model.script.Regel Horizon] - Command sent to Horizon power plug is OFF!
`

Great! So maybe we could ask to merge the fix with the source code? How does this work? Do we have to inform somebody?

Stupid that I don't know why the manual add-on didn't want at the beginning. Strange...

--- a/bundles/binding/org.openhab.binding.caldav-personal/src/main/java/org/openhab/binding/caldav_personal/internal/CalDavBinding.java
+++ b/bundles/binding/org.openhab.binding.caldav-personal/src/main/java/org/openhab/binding/caldav_personal/internal/CalDavBinding.java
@@ -167,12 +167,18 @@ public class CalDavBinding extends AbstractBinding<CalDavBindingProvider> implem
      */
     @Override
     public void eventBegins(CalDavEvent event) {
+        logger.debug("eventBegins() called for event '{}'", event.getShortName());
+
         if (!calendars.contains(event.getCalendarId())) {
+            logger.trace("event calendarId does not exist in calendars");
             return;
         }
 
+        logger.debug("check start of event '{}'", event.getShortName());
         if (event.getStart().isBeforeNow()) {
-            return;
+            logger.debug("event is before now; ignoring and updating anyway.");
+            logger.trace("event start time: {} ", event.getStart());
+            logger.trace("now:              {} ", DateTime.now());
         }
 
         logger.debug("adding event to map: {}", event.getShortName());

